certdepot/impl.go

```diff
@@ -0,0 +1,180 @@
+package certdepot
+
+import (
+	"time"
+
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"github.com/square/certstrap/depot"
+	mgo "gopkg.in/mgo.v2"
+)
+
+type DBUser struct {
+	ID   string    `bson:"_id"`
+	Cert string    `bson:"cert"`
+	TTL  time.Time `bson:"ttl"`
+}
+
+var (
+	dbUserIDKey   = bsonutil.MustHaveTag(DBUser{}, "ID")
+	dbUserCertKey = bsonutil.MustHaveTag(DBUser{}, "Cert")
+	dbUserTTLKey  = bsonutil.MustHaveTag(DBUser{}, "TTL")
+)
+
+type mongoCertDepot struct {
+	session        *mgo.Session
+	databaseName   string
+	collectionName string
+	expireAfter    time.Duration
+}
+
+type MgoCertDepotOptions struct {
+	MongoDBURI           string
+	MongoDBDialTimeout   time.Duration
+	MongoDBSocketTimeout time.Duration
+	DatabaseName         string
+	CollectionName       string
+	ExpireAfter          time.Duration
+}
+
+// Create a new cert depot in the specified MongoDB.
+func NewMgoCertDepot(opts MgoCertDepotOptions) (depot.Depot, error) {
+	opts = defaults(opts)
+
+	session, err := mgo.DialWithTimeout(opts.MongoDBURI, opts.MongoDBDialTimeout)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not connect to db %s", opts.MongoDBURI)
+	}
+	session.SetSocketTimeout(opts.MongoDBSocketTimeout)
+
+	return &mongoCertDepot{
+		session:        session,
+		databaseName:   opts.DatabaseName,
+		collectionName: opts.CollectionName,
+		expireAfter:    opts.ExpireAfter,
+	}, nil
+}
+
+// Create a new cert depot in the specified MongoDB, using an existing session.
+func NewMgoCertDepotWithSession(s *mgo.Session, opts MgoCertDepotOptions) (depot.Depot, error) {
+	if s == nil {
+		return NewMgoCertDepot(opts)
+	}
+
+	opts = defaults(opts)
+	return &mongoCertDepot{
+		session:        s,
+		databaseName:   opts.DatabaseName,
+		collectionName: opts.CollectionName,
+		expireAfter:    opts.ExpireAfter,
+	}, nil
+}
```

Comment:
I'd refactor line 52-57 and 67-72 into a private function. 

also see if we want to do any option validation? is there a possibility to error if the options are not valid?

Reply:
error validation done in defaults (I can change the name to validate)